File: tests/run_solutions.py
import subprocess
import os

for year in range(2015, 2024):
  for root, dirs, files in os.walk(os.path.join('..', f'{year}')):
    for file in files:
      if 'solution' in file:
        solution_file = os.path.join(root, file)
        try:
          if 'js' in file:
            print(f'Running {solution_file}')
            result = subprocess.run(['node', solution_file], check=True, text=True, capture_output=True)
          # elif 'go' in file:
          #   result = subprocess.run(['go run', solution_file], check=True, text=True, capture_output=True)
          # elif 'py' in file:
          #   print(f'Running {solution_file}')
          #   result = subprocess.run(['py', solution_file], check=True, text=True, capture_output=True)
          else:
            continue

          stdout_output = result.stdout
          stderr_output = result.stderr

          if stdout_output:
              print('Standard Output:')
              print(stdout_output)

          if stderr_output:
              print('Standard Error:')
              print(stderr_output)

        except subprocess.CalledProcessError as e:
          print(f'An error occurred: {e}')

File: tests/runSolutions.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

func main() {
	for year := 2015; year < 2024; year++ {
		dirPath := filepath.Join("..", fmt.Sprintf("%d", year))

		err := filepath.Walk(dirPath, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}

			if !info.IsDir() && strings.Contains(info.Name(), "solution") {
				switch {
				case strings.HasSuffix(info.Name(), ".js"):
					fmt.Printf("Running %s\n", path)
					cmd := exec.Command("node", path)
					runCommand(cmd)
				case strings.HasSuffix(info.Name(), ".go"):
					fmt.Printf("Running %s\n", path)
					cmd := exec.Command("go", "run", path)
					runCommand(cmd)
					// case strings.HasSuffix(info.Name(), ".py"):
					// 	fmt.Printf("Running %s\n", path)
					// 	cmd := exec.Command("python3", path)
					// 	runCommand(cmd)
				}
			}
			return nil
		})

		if err != nil {
			fmt.Println("Error:", err)
		}
	}
}

func runCommand(cmd *exec.Cmd) {
	output, err := cmd.CombinedOutput()
	if err != nil {
		fmt.Printf("An error occurred: %s\n", err)
	}
	if len(output) > 0 {
		fmt.Println("Output:")
		fmt.Println(string(output))
	}
}
